File: selenium-tests/excelReporter.js
import ExcelJS from 'exceljs';
import path from 'path';

/**
 * Generates a color-coded E2E test report in Excel format.
 * @param {Array} results - List of test case result objects
 * @param {string} outputPath - Output file path
 */
export async function generateReport(results, outputPath) {
  const workbook = new ExcelJS.Workbook();
  workbook.creator = 'Smart Ambulance E2E Test Suite';
  workbook.lastModifiedBy = 'Smart Ambulance E2E Test Suite';
  workbook.created = new Date();
  workbook.modified = new Date();

  const worksheet = workbook.addWorksheet('E2E Test Results');

  // Ensure gridlines are visible
  worksheet.views = [{ showGridLines: true }];

  // Calculate statistics
  const total = results.length;
  const passed = results.filter(r => r.status === 'PASS').length;
  const failed = total - passed;
  const passRate = total > 0 ? ((passed / total) * 100).toFixed(1) + '%' : '0%';

  // --- Title Block ---
  worksheet.mergeCells('A1:H1');
  const titleCell = worksheet.getCell('A1');
  titleCell.value = 'Smart Ambulance E2E Test Suite - Complete E2E Report';
  titleCell.font = { name: 'Arial', size: 16, bold: true, color: { argb: 'FFFFFFFF' } };
  titleCell.fill = {
    type: 'pattern',
    pattern: 'solid',
    fgColor: { argb: 'FF1E293B' } // Slate 800
  };
  titleCell.alignment = { horizontal: 'center', vertical: 'middle' };
  worksheet.getRow(1).height = 40;

  // --- Summary Dashboard (Cards) ---
  // Total Cases Card
  worksheet.mergeCells('A3:B4');
  const totalCard = worksheet.getCell('A3');
  totalCard.value = `TOTAL CASES\n\n${total}`;
  totalCard.font = { name: 'Arial', size: 10, bold: true, color: { argb: 'FF334155' } };
  totalCard.alignment = { horizontal: 'center', vertical: 'middle', wrapText: true };
  totalCard.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FFF1F5F9' } }; // Slate 100
  setBorder(worksheet, 'A3:B4', 'FFCBD5E1');

  // Passed Card
  worksheet.mergeCells('C3:D4');
  const passedCard = worksheet.getCell('C3');
  passedCard.value = `PASSED\n\n${passed}`;
  passedCard.font = { name: 'Arial', size: 10, bold: true, color: { argb: 'FF15803D' } }; // Green 700
  passedCard.alignment = { horizontal: 'center', vertical: 'middle', wrapText: true };
  passedCard.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FFDCFCE7' } }; // Green 100
  setBorder(worksheet, 'C3:D4', 'FF86EFAC');

  // Failed Card
  worksheet.mergeCells('E3:F4');
  const failedCard = worksheet.getCell('E3');
  failedCard.value = `FAILED\n\n${failed}`;
  failedCard.font = { name: 'Arial', size: 10, bold: true, color: { argb: 'FFB91C1C' } }; // Red 700
  failedCard.alignment = { horizontal: 'center', vertical: 'middle', wrapText: true };
  failedCard.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FFFEE2E2' } }; // Red 100
  setBorder(worksheet, 'E3:F4', 'FFFCA5A5');

  // Pass Rate Card
  worksheet.mergeCells('G3:H4');
  const rateCard = worksheet.getCell('G3');
  rateCard.value = `PASS RATE\n\n${passRate}`;
  rateCard.font = { name: 'Arial', size: 10, bold: true, color: { argb: 'FF1D4ED8' } }; // Blue 700
  rateCard.alignment = { horizontal: 'center', vertical: 'middle', wrapText: true };
  rateCard.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FFDBEAFE' } }; // Blue 100
  setBorder(worksheet, 'G3:H4', 'FF93C5FD');

  worksheet.getRow(3).height = 20;
  worksheet.getRow(4).height = 20;

  // --- Spacer Row ---
  worksheet.getRow(5).height = 15;

  // --- Table Headers ---
  const headers = [
    'Case ID',
    'Test Suite',
    'Test Scenario / Description',
    'Test Inputs',
    'Expected Behavior',
    'Actual Outcome',
    'Status',
    'Duration (ms)'
  ];

  const headerRowNumber = 6;
  const headerRow = worksheet.getRow(headerRowNumber);
  headerRow.values = headers;
  headerRow.height = 28;

  headers.forEach((h, idx) => {
    const cell = headerRow.getCell(idx + 1);
    cell.font = { name: 'Arial', size: 11, bold: true, color: { argb: 'FFFFFFFF' } };
    cell.fill = {
      type: 'pattern',
      pattern: 'solid',
      fgColor: { argb: 'FF475569' } // Slate 600
    };
    cell.alignment = { horizontal: 'center', vertical: 'middle', wrapText: true };
    cell.border = {
      top: { style: 'medium', color: { argb: 'FF334155' } },
      bottom: { style: 'medium', color: { argb: 'FF334155' } },
      left: { style: 'thin', color: { argb: 'FFCBD5E1' } },
      right: { style: 'thin', color: { argb: 'FFCBD5E1' } }
    };
  });

  // --- Data Rows ---
  let currentRowNumber = 7;
  results.forEach(result => {
    const row = worksheet.getRow(currentRowNumber);
    row.values = [
      result.id,
      result.suite,
      result.description,
      result.inputs,
      result.expected,
      result.actual,
      result.status,
      result.durationMs
    ];
    row.height = 22;

    // Apply color schemes based on Status (PASS / FAIL)
    const isPass = result.status === 'PASS';
    const cellBgColor = isPass ? 'FFF0FDF4' : 'FFFDF2F2'; // Very light green / red for rows
    const statusTextClr = isPass ? 'FF15803D' : 'FFB91C1C';
    const statusBgClr = isPass ? 'FFDCFCE7' : 'FFFEE2E2';

    for (let colIdx = 1; colIdx <= 8; colIdx++) {
      const cell = row.getCell(colIdx);
      cell.font = { name: 'Arial', size: 10, color: { argb: 'FF1E293B' } };
      cell.border = {
        top: { style: 'thin', color: { argb: 'FFE2E8F0' } },
        bottom: { style: 'thin', color: { argb: 'FFE2E8F0' } },
        left: { style: 'thin', color: { argb: 'FFE2E8F0' } },
        right: { style: 'thin', color: { argb: 'FFE2E8F0' } }
      };

      // Set cell fills
      if (colIdx === 7) {
        // Status column has unique bubble/pill look
        cell.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: statusBgClr } };
        cell.font = { name: 'Arial', size: 10, bold: true, color: { argb: statusTextClr } };
        cell.alignment = { horizontal: 'center', vertical: 'middle' };
      } else {
        cell.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: cellBgColor } };
        
        // Alignment formatting
        if (colIdx === 1 || colIdx === 8) {
          cell.alignment = { horizontal: 'center', vertical: 'middle' };
        } else if (colIdx === 2) {
          cell.alignment = { horizontal: 'center', vertical: 'middle' };
        } else {
          cell.alignment = { horizontal: 'left', vertical: 'middle', wrapText: true };
        }
      }
    }
    currentRowNumber++;
  });

  // Enable Auto Filter
  worksheet.autoFilter = {
    from: { row: headerRowNumber, column: 1 },
    to: { row: currentRowNumber - 1, column: 8 }
  };

  // Adjust columns to fit content
  worksheet.columns.forEach((column, idx) => {
    let maxLen = 0;
    // Don't size columns based on the merged title row
    column.eachCell({ includeEmpty: false }, (cell, rowNum) => {
      if (rowNum >= headerRowNumber) {
        const valStr = cell.value ? cell.value.toString() : '';
        if (valStr.length > maxLen) {
          maxLen = valStr.length;
        }
      }
    });

    // Limit maximum size to prevent huge wraps, set reasonable defaults
    let width = Math.max(maxLen + 4, 12);
    if (idx === 0) width = 10;  // Case ID
    if (idx === 1) width = 15;  // Test Suite
    if (idx === 2) width = 35;  // Scenario
    if (idx === 3) width = 30;  // Inputs
    if (idx === 4) width = 35;  // Expected
    if (idx === 5) width = 35;  // Actual
    if (idx === 6) width = 12;  // Status
    if (idx === 7) width = 14;  // Duration (ms)

    column.width = width;
  });

  // Save the report
  await workbook.xlsx.writeFile(outputPath);
  console.log(`[ExcelReporter] Excel E2E Test Report saved to: ${outputPath}`);
}

// Helper function to set borders on cell range
function setBorder(worksheet, range, borderColor) {
  const [start, end] = range.split(':');
  const startCol = start.charCodeAt(0) - 64;
  const startRow = parseInt(start.slice(1));
  const endCol = end.charCodeAt(0) - 64;
  const endRow = parseInt(end.slice(1));

  for (let r = startRow; r <= endRow; r++) {
    for (let c = startCol; c <= endCol; c++) {
      const cell = worksheet.getCell(r, c);
      cell.border = {
        top: { style: 'thin', color: { argb: borderColor } },
        bottom: { style: 'thin', color: { argb: borderColor } },
        left: { style: 'thin', color: { argb: borderColor } },
        right: { style: 'thin', color: { argb: borderColor } }
      };
    }
  }
}

File: selenium-tests/excel_reporter.py
import os
import time
from datetime import datetime
import openpyxl
from openpyxl.styles import Font, PatternFill, Alignment, Border, Side
from openpyxl.utils import get_column_letter
from openpyxl.chart import PieChart, Reference

class ExcelReporter:
    def __init__(self, filename="Selenium_Web_300_Test_Report.xlsx"):
        self.filename = filename
        self.wb = openpyxl.Workbook()
        
    def generate_report(self, test_results, start_time, end_time):
        total_tests = len(test_results)
        passed = sum(1 for r in test_results if r['status'] == 'PASS')
        failed = sum(1 for r in test_results if r['status'] == 'FAIL')
        skipped = sum(1 for r in test_results if r['status'] == 'SKIP')
        pass_rate = (passed / total_tests * 100) if total_tests > 0 else 0.0
        total_duration = end_time - start_time

        # --- Sheet 1: Executive Dashboard ---
        ws_dash = self.wb.active
        ws_dash.title = "Executive Dashboard"
        ws_dash.views.sheetView[0].showGridLines = True

        # Header Title Banner
        ws_dash.merge_cells("A1:G2")
        title_cell = ws_dash["A1"]
        title_cell.value = "Smart Ambulance System - Selenium Web Automation Test Report (300 Test Cases)"
        title_cell.font = Font(name="Calibri", size=16, bold=True, color="FFFFFF")
        title_cell.fill = PatternFill(start_color="1E293B", end_color="1E293B", fill_type="solid")
        title_cell.alignment = Alignment(horizontal="center", vertical="center")

        # Summary Metrics Cards
        headers_metrics = [
            ("Total Test Cases", total_tests, "388E3C"),
            ("Passed Test Cases", passed, "2E7D32"),
            ("Failed Test Cases", failed, "D32F2F" if failed > 0 else "757575"),
            ("Skipped Test Cases", skipped, "F57C00" if skipped > 0 else "757575"),
            ("Pass Rate", f"{pass_rate:.2f}%", "1B5E20"),
            ("Total Duration", f"{total_duration:.2f} s", "0288D1")
        ]

        row_start = 4
        for idx, (label, val, color) in enumerate(headers_metrics, start=1):
            col_letter = get_column_letter(idx)
            cell_lbl = ws_dash[f"{col_letter}{row_start}"]
            cell_lbl.value = label
            cell_lbl.font = Font(name="Calibri", size=10, bold=True, color="475569")
            cell_lbl.alignment = Alignment(horizontal="center", vertical="center")
            cell_lbl.fill = PatternFill(start_color="F1F5F9", end_color="F1F5F9", fill_type="solid")

            cell_val = ws_dash[f"{col_letter}{row_start+1}"]
            cell_val.value = val
            cell_val.font = Font(name="Calibri", size=14, bold=True, color=color)
            cell_val.alignment = Alignment(horizontal="center", vertical="center")
            cell_val.fill = PatternFill(start_color="FFFFFF", end_color="FFFFFF", fill_type="solid")
            
            thin_border = Border(
                left=Side(style='thin', color='CBD5E1'),
                right=Side(style='thin', color='CBD5E1'),
                top=Side(style='thin', color='CBD5E1'),
                bottom=Side(style='thin', color='CBD5E1')
            )
            cell_lbl.border = thin_border
            cell_val.border = thin_border

        # Module Breakdown Table
        ws_dash["A7"] = "Module-wise Test Breakdown"
        ws_dash["A7"].font = Font(name="Calibri", size=12, bold=True, color="0F172A")

        table_headers = ["Module Name", "Total Tests", "Passed", "Failed", "Skipped", "Pass Rate"]
        for col_idx, th in enumerate(table_headers, start=1):
            cell = ws_dash.cell(row=8, column=col_idx, value=th)
            cell.font = Font(name="Calibri", size=10, bold=True, color="FFFFFF")
            cell.fill = PatternFill(start_color="334155", end_color="334155", fill_type="solid")
            cell.alignment = Alignment(horizontal="center", vertical="center")

        modules = {}
        for r in test_results:
            m = r['module']
            if m not in modules:
                modules[m] = {'total': 0, 'pass': 0, 'fail': 0, 'skip': 0}
            modules[m]['total'] += 1
            if r['status'] == 'PASS':
                modules[m]['pass'] += 1
            elif r['status'] == 'FAIL':
                modules[m]['fail'] += 1
            else:
                modules[m]['skip'] += 1

        curr_row = 9
        for mod_name, data in modules.items():
            mod_pass_rate = (data['pass'] / data['total'] * 100) if data['total'] > 0 else 0.0
            ws_dash.cell(row=curr_row, column=1, value=mod_name).alignment = Alignment(horizontal="left")
            ws_dash.cell(row=curr_row, column=2, value=data['total']).alignment = Alignment(horizontal="center")
            ws_dash.cell(row=curr_row, column=3, value=data['pass']).alignment = Alignment(horizontal="center")
            ws_dash.cell(row=curr_row, column=4, value=data['fail']).alignment = Alignment(horizontal="center")
            ws_dash.cell(row=curr_row, column=5, value=data['skip']).alignment = Alignment(horizontal="center")
            
            rate_cell = ws_dash.cell(row=curr_row, column=6, value=f"{mod_pass_rate:.2f}%")
            rate_cell.alignment = Alignment(horizontal="center")
            rate_cell.font = Font(bold=True, color="166534" if mod_pass_rate == 100 else "991B1B")
            
            for c in range(1, 7):
                ws_dash.cell(row=curr_row, column=c).border = Border(
                    left=Side(style='thin', color='E2E8F0'),
                    right=Side(style='thin', color='E2E8F0'),
                    top=Side(style='thin', color='E2E8F0'),
                    bottom=Side(style='thin', color='E2E8F0')
                )
            curr_row += 1

        # Adjust Column Widths for Dashboard
        for col in ws_dash.columns:
            max_len = max(len(str(cell.value or '')) for cell in col)
            col_letter = get_column_letter(col[0].column)
            ws_dash.column_dimensions[col_letter].width = max(max_len + 4, 15)

        # --- Sheet 2: Test Execution Details ---
        ws_details = self.wb.create_sheet(title="Test Execution Details")
        ws_details.views.sheetView[0].showGridLines = True

        detail_headers = [
            "Test ID", "Module / Suite", "Test Name", "Description", 
            "Status", "Duration (s)", "Timestamp", "Assertion / Log Detail"
        ]
        
        ws_details.append(detail_headers)
        for col_num in range(1, len(detail_headers) + 1):
            cell = ws_details.cell(row=1, column=col_num)
            cell.font = Font(name="Calibri", size=11, bold=True, color="FFFFFF")
            cell.fill = PatternFill(start_color="1E293B", end_color="1E293B", fill_type="solid")
            cell.alignment = Alignment(horizontal="center", vertical="center")

        pass_fill = PatternFill(start_color="DCFCE7", end_color="DCFCE7", fill_type="solid")
        pass_font = Font(color="15803D", bold=True)
        
        fail_fill = PatternFill(start_color="FEE2E2", end_color="FEE2E2", fill_type="solid")
        fail_font = Font(color="B91C1C", bold=True)

        for row_idx, res in enumerate(test_results, start=2):
            ws_details.append([
                res['id'],
                res['module'],
                res['name'],
                res['description'],
                res['status'],
                round(res['duration'], 4),
                res['timestamp'],
                res['details']
            ])

            status_cell = ws_details.cell(row=row_idx, column=5)
            if res['status'] == 'PASS':
                status_cell.fill = pass_fill
                status_cell.font = pass_font
            else:
                status_cell.fill = fail_fill
                status_cell.font = fail_font

            ws_details.cell(row=row_idx, column=1).alignment = Alignment(horizontal="center")
            ws_details.cell(row=row_idx, column=5).alignment = Alignment(horizontal="center")
            ws_details.cell(row=row_idx, column=6).alignment = Alignment(horizontal="right")
            ws_details.cell(row=row_idx, column=7).alignment = Alignment(horizontal="center")

            for col_idx in range(1, 9):
                ws_details.cell(row=row_idx, column=col_idx).border = Border(
                    left=Side(style='thin', color='F1F5F9'),
                    right=Side(style='thin', color='F1F5F9'),
                    top=Side(style='thin', color='F1F5F9'),
                    bottom=Side(style='thin', color='F1F5F9')
                )

        # Freeze headers on details sheet
        ws_details.freeze_panes = "A2"

        # Adjust Column Widths for Details
        for col in ws_details.columns:
            max_len = max(len(str(cell.value or '')) for cell in col)
            col_letter = get_column_letter(col[0].column)
            ws_details.column_dimensions[col_letter].width = min(max(max_len + 3, 12), 60)

        # Save workbook
        os.makedirs(os.path.dirname(os.path.abspath(self.filename)), exist_ok=True)
        self.wb.save(self.filename)
        print(f"[ExcelReporter] Report successfully saved to '{self.filename}' with {total_tests} test cases.")
